New vinscan

simsong opened the pull request.

kamwoods committed:
Initial commit of Vehicle Identification Number (VIN) scanner
New files associated with this scanner are:

- scan_vehicle.flex: A pattern scanner to recognize vehicle related
                     identifiers. Currently limited to VINs, but
                     could be expanded to include HINs, etc.
                     Patterned after scan_accts.flex

- scan_vin.cpp:      A validator for Vehicle Identification Numbers
                     (VINs). Validates rules for World Manufacturer
                     Identifier, Vehicle Descriptor Section (including
                     check digit) and Vehicle Identifier Section to
                     ensure capture of valid VINs. Structure adapted
                     from scan_ccns2.cpp

- scan_vin.h:        Header for scan_vin.cpp

Some initial test files are also included, using synthesized valid
and invalid VINs:

- src/tests/test_synthetic_vin.json
- src/tests/test_vin.txt
- src/tests/test_vin_doc.odt

Build and scanner registration files modified:

- src/Makefile.am
- src/bulk_extractor_scanners.h

kamwoods committed:
Unit tests for VIN scanner (scan_vin.cpp)

Individual scanner functionality tests added to test_be1 for scan_vin
scanner to validate operation, consistent with other test structures.

Included tests:
- scan_vin_validation: general tests for valid and invalid VINs
- scan_vin1:           test with valid VINs labeled in context
- scan_vin2:           test with VINs in open text, no context
- scan_vin3:           invalid VIN tests (disallowed characters, bad check
                       digits, lowercase characters, all-digit VINs)
- scan_vin_json:       scan valid/invalid VINs appearing in JSON file
- scan_vin_year_codes: test year codes
- scan_vin_context:    test that VINs in hex dump context are disregarded

Verified via: make test_be && test_be "[scanners]"

kamwoods committed:
Additional tests for VIN scanner

Integration tests added to test_be2 for scan_vehicle to validate operation,
consistent with existing test structures.

Included tests:

- test_vin: test with test_vin.txt from the tests/ directory
- test_vin_json: test with test_synthetic_vin.json from the tests/ directory
- test_vin_odt: test with the ODT doc in the tests/ directory

Verified via: make test_be && test_be "[phase1]"

simsong commented:
@kamwoods — Any idea why codecov didn't run here?

kamwoods committed:
Resolved some issues identified by Copilot

- Error suppresion logic in scan_vehicle.flex
- Character check bug in scan_vin.cpp
- Misc comment typos
- Removed an unused pattern from scan_vehicle.flex

simsong reviewed (commented):
looks great.

kamwoods commented on src/scan_vin.cpp:
```diff
@@ -0,0 +1,278 @@
+    // J-R: Asia
+    // S-Z: Europe
+    
+    if (strchr("0", first)) return -1;  // 0 is not used as first character
```
Resolved in latest commit to new-vinscan.

kamwoods committed:
Code cleanup

- Removed dead code blocks from a previous approach in scan_vin.cpp
  and scan_vin.h

kamwoods commented:
Cleaned up some dead code blocks after a codecov run. Coverage on scan_accts.flex and scan_vin.cpp improved.

kamwoods committed:
VIN histogram cleanup

Minor modification of scan_vehicle.flex to stop generating
extraneous vin_manufacturer.txt and vin_year.txt histograms.

Brings the vehicle scan histogram generation behavior in line
with other bulk_extractor scanners, generating a single
vin_histogram.txt on each run.
